exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/SpanAdapter.java

```diff
@@ -156,23 +194,38 @@ static Span toProtoSpan(SpanData spanData) {
     return builder.build();
   }
 
+  // Visible for testing
   static Status toStatusProto(StatusData status) {
-    Status.StatusCode protoStatusCode = Status.StatusCode.STATUS_CODE_UNSET;
-    Status.DeprecatedStatusCode deprecatedStatusCode = DEPRECATED_STATUS_CODE_OK;
-    if (status.getStatusCode() == StatusCode.OK) {
-      protoStatusCode = Status.StatusCode.STATUS_CODE_OK;
-    } else if (status.getStatusCode() == StatusCode.ERROR) {
-      protoStatusCode = Status.StatusCode.STATUS_CODE_ERROR;
-      deprecatedStatusCode = DEPRECATED_STATUS_CODE_UNKNOWN_ERROR;
+    final Status withoutDescription;
+    switch (status.getStatusCode()) {
+      case OK:
+        withoutDescription = STATUS_OK;
+        break;
+      case ERROR:
+        withoutDescription = STATUS_ERROR;
+        break;
+      case UNSET:
+      default:
+        withoutDescription = STATUS_UNSET;
+        break;
     }
+    if (status.getDescription().isEmpty()) {
+      return withoutDescription;
+    }
+    return withoutDescription.toBuilder().setMessage(status.getDescription()).build();
+  }
 
-    @SuppressWarnings("deprecation") // setDeprecatedCode is deprecated.
-    Status.Builder builder =
-        Status.newBuilder().setCode(protoStatusCode).setDeprecatedCode(deprecatedStatusCode);
-    if (!status.getDescription().isEmpty()) {
-      builder.setMessage(status.getDescription());
+  private static ThreadLocalCache getThreadLocalCache() {
+    ThreadLocalCache result = THREAD_LOCAL_CACHE.get();
+    if (result == null) {
+      result = new ThreadLocalCache();
+      THREAD_LOCAL_CACHE.set(result);
     }
-    return builder.build();
+    return result;
+  }
+
+  static final class ThreadLocalCache {
```

Comment:
Extendable to cache any and all buffers :) Think I'm done for now, but I actually had good result caching `Span.Builder`. I didn't push it though since I couldn't reason about it when looking at the code, I didn't see proto reusing any `Collection`s between `build()` and `clear()` so I was surprised to see another several hundred bytes of allocations saved. Anyways such an optimization wouldn't matter when switching to direct marshalling, while the others generally should.